rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -1531,103 +1529,51 @@ impl TryFrom<(pb_queues::CanisterQueues, &dyn CheckpointLoadingMetrics)> for Can
     fn try_from(
         (item, metrics): (pb_queues::CanisterQueues, &dyn CheckpointLoadingMetrics),
     ) -> Result<Self, Self::Error> {
-        let mut canister_queues = BTreeMap::new();
-        let mut pool = MessagePool::default();
-        let mut shed_responses = BTreeMap::new();
-
-        if !item.input_queues.is_empty() || !item.output_queues.is_empty() {
-            // Backward compatibility: deserialize from `input_queues` and `output_queues`.
-
-            if item.pool.is_some() || !item.canister_queues.is_empty() {
-                return Err(ProxyDecodeError::Other(
-                    "Both `input_queues`/`output_queues` and `pool`/`canister_queues` are populated"
-                        .to_string(),
-                ));
-            }
-
-            if item.input_queues.len() != item.output_queues.len() {
-                return Err(ProxyDecodeError::Other(format!(
-                    "CanisterQueues: Mismatched input ({}) and output ({}) queue lengths",
-                    item.input_queues.len(),
-                    item.output_queues.len()
-                )));
-            }
-            for (ie, oe) in item
-                .input_queues
-                .into_iter()
-                .zip(item.output_queues.into_iter())
-            {
-                if ie.canister_id != oe.canister_id {
-                    return Err(ProxyDecodeError::Other(format!(
-                        "CanisterQueues: Mismatched input {:?} and output {:?} queue entries",
-                        ie.canister_id, oe.canister_id
-                    )));
-                }
+        let pool = MessagePool::try_from(item.pool.unwrap_or_default())?;
+        let shed_responses = item
+            .shed_responses
+            .into_iter()
+            .map(|sr| {
+                let sr = message_pool::CallbackReference::try_from(sr)?;
+                Ok((sr.0, sr.1))
```

Comment:
I didn't look into any detail but can't you write this in one line? i.e. `Ok(message_pool::CallbackReference::try_from(sr)?)`

Reply:
That doesn't work. `sr` here is a `CallbackReference` newtype, not a tuple (because Rust won't let me implement conversion betwee two types, proto and tuple, in what's essentially a third crate).